File: packages/runtime/nodejs/src/workflow/aggregate/document/flat-schema.ts
import { FlowGramNode, WorkflowNodeSchema, WorkflowSchema } from '@flowgram.ai/runtime-interface';

import { WorkflowRuntimeEdge } from '@workflow/entity';

export interface FlattenData {
  flattenSchema: WorkflowSchema;
  nodeBlocks: Map<string, string[]>;
  nodeEdges: Map<string, string[]>;
}

type FlatSchema = (json: Partial<WorkflowSchema>) => FlattenData;

const flatLayer = (data: FlattenData, layerNodes: WorkflowNodeSchema[]) => {
  layerNodes.forEach((nodeJSON) => {
    const { blocks, edges } = nodeJSON;
    if (blocks) {
      data.flattenSchema.nodes.push(...blocks);
      const blockIDs: string[] = [];
      blocks.forEach((block) => {
        if (block.blocks) {
          flatLayer(data, block.blocks);
        }
        blockIDs.push(block.id);
      });
      data.nodeBlocks.set(nodeJSON.id, blockIDs);
      delete nodeJSON.blocks;
    }
    if (edges) {
      data.flattenSchema.edges.push(...edges);
      const edgeIDs: string[] = [];
      edges.forEach((edge) => {
        const edgeID = WorkflowRuntimeEdge.createID(edge);
        edgeIDs.push(edgeID);
      });
      data.nodeEdges.set(nodeJSON.id, edgeIDs);
      delete nodeJSON.edges;
    }
  });
};

/**
 * flat the tree json structure, extract the structure information to map
 */
export const flatSchema: FlatSchema = (schema = { nodes: [], edges: [] }) => {
  const rootNodes = schema.nodes ?? [];
  const rootEdges = schema.edges ?? [];

  const data: FlattenData = {
    flattenSchema: {
      nodes: [...rootNodes],
      edges: [...rootEdges],
    },
    nodeBlocks: new Map(),
    nodeEdges: new Map(),
  };

  const rootBlockIDs: string[] = rootNodes.map((node) => node.id);
  const rootEdgeIDs: string[] = rootEdges.map((edge) => WorkflowRuntimeEdge.createID(edge));

  data.nodeBlocks.set(FlowGramNode.Root, rootBlockIDs);
  data.nodeEdges.set(FlowGramNode.Root, rootEdgeIDs);

  flatLayer(data, rootNodes);

  return data;
};

File: packages/runtime-go/domain/document/flat-schema.go
package document

import (
	"flowgram-runtime-go/interface/constant"
	schemaType "flowgram-runtime-go/interface/schema"
)

// FlattenData represents the flattened workflow schema data
type FlattenData struct {
	FlattenSchema *schemaType.WorkflowSchema
	NodeBlocks    map[string][]string
	NodeEdges     map[string][]string
}

func flatLayer(data *FlattenData, nodeSchema *schemaType.WorkflowNodeSchema) {
	if nodeSchema.Blocks != nil {
		data.FlattenSchema.Nodes = append(data.FlattenSchema.Nodes, nodeSchema.Blocks...)
		blockIDs := make([]string, 0)

		for _, block := range nodeSchema.Blocks {
			blockIDs = append(blockIDs, block.ID)
			// Recursively process child nodes' blocks and edges
			if block.Blocks != nil {
				flatLayer(data, &block)
			}
		}

		data.NodeBlocks[nodeSchema.ID] = blockIDs
		nodeSchema.Blocks = nil
	}

	if nodeSchema.Edges != nil {
		data.FlattenSchema.Edges = append(data.FlattenSchema.Edges, nodeSchema.Edges...)
		edgeIDs := make([]string, 0)

		for _, edge := range nodeSchema.Edges {
			edgeID := WorkflowRuntimeEdge_CreateID(&edge)
			edgeIDs = append(edgeIDs, edgeID)
		}

		data.NodeEdges[nodeSchema.ID] = edgeIDs
		nodeSchema.Edges = nil
	}
}

// FlatSchema flattens the tree json structure and extracts the structure information to maps
func FlatSchema(schema *schemaType.WorkflowSchema) *FlattenData {
	if schema == nil {
		schema = &schemaType.WorkflowSchema{
			Nodes: make([]schemaType.WorkflowNodeSchema, 0),
			Edges: make([]schemaType.WorkflowEdgeSchema, 0),
		}
	}

	rootNodes := schema.Nodes
	rootEdges := schema.Edges

	data := &FlattenData{
		FlattenSchema: &schemaType.WorkflowSchema{
			Nodes: make([]schemaType.WorkflowNodeSchema, 0),
			Edges: make([]schemaType.WorkflowEdgeSchema, 0),
		},
		NodeBlocks: make(map[string][]string),
		NodeEdges:  make(map[string][]string),
	}

	root := &schemaType.WorkflowNodeSchema{
		ID:     string(constant.FlowGramNodeRoot),
		Type:   constant.FlowGramNodeRoot,
		Blocks: rootNodes,
		Edges:  rootEdges,
		Meta: schemaType.WorkflowNodeMetaSchema{
			Position: schemaType.PositionSchema{
				X: 0,
				Y: 0,
			},
		},
		Data: make(map[string]interface{}),
	}

	flatLayer(data, root)

	return data
}
